### PR TITLE
[SPARK-32777][SQL] Aggregation support aggregate function with multiple foldable expressions.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Spark SQL exists a bug show below:

```
spark.sql(
  " SELECT COUNT(DISTINCT 2), COUNT(DISTINCT 2, 3)")
  .show()
+-----------------+--------------------+
|count(DISTINCT 2)|count(DISTINCT 2, 3)|
+-----------------+--------------------+
|                1|                   1|
+-----------------+--------------------+

spark.sql(
  " SELECT COUNT(DISTINCT 2), COUNT(DISTINCT 3, 2)")
  .show()
+-----------------+--------------------+
|count(DISTINCT 2)|count(DISTINCT 3, 2)|
+-----------------+--------------------+
|                1|                   0|
+-----------------+--------------------+
```
The first query is correct, but the second query is not.
The root reason is the second query rewrited by `RewriteDistinctAggregates` who expand the output but lost the 2.

### Why are the changes needed?
Fix a bug.
`SELECT COUNT(DISTINCT 2), COUNT(DISTINCT 3, 2)` should return `1, 1`

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
New UT

### CODE REVIEW

**Comment by beliefer** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @cloud-fan and @linhongliu-db

File: sql/core/src/test/resources/sql-tests/results/group-by-filter.sql.out [hunk 1]
```diff
@@ -150,6 +150,38 @@ struct<count(DISTINCT id) FILTER (WHERE false):bigint>
...
+-- !query
+SELECT COUNT(DISTINCT id), COUNT(DISTINCT 3,2) FILTER (WHERE dept_id = 40) FROM emp
+-- !query schema
+struct<count(DISTINCT id):bigint,count(DISTINCT 3, 2) FILTER (WHERE (dept_id = 40)):bigint>
+-- !query output
+8	0
+
+
+-- !query
+SELECT COUNT(DISTINCT id), COUNT(DISTINCT 2,3) FILTER (WHERE dept_id = 30) FROM emp
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/group-by-filter.sql.out:170]: can we test `WHERE dept_id > 0`? to make sure that we get correct result even if there are more than one inputs after filter.

Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/results/group-by-filter.sql.out:170]: OK

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 2]
```diff
@@ -216,20 +216,24 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
...
+        // we must expand at least one of the children (here we take the first child),
+        // or If we don't, we will get the wrong result, for example:
+        // count(distinct 1) will be explained to count(1) after the rewrite function.
+        // Generally, the distinct aggregateFunction should not run
+        // foldable TypeCheck for the first child.
+        e -> e.aggregateFunction.children.take(1).toSet
+      }
+    }
+    val distinctAggGroupLookup = distinctAggGroupMap.toMap
+    val distinctAggGroups = distinctAggGroupMap.groupBy(_._2).map{ kv =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala:235]: nit: `mapValues`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 3]
```diff
@@ -293,12 +297,16 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
           val operators = expressions.map { e =>
             val af = e.aggregateFunction
             val naf = patchAggregateFunctionChildren(af) { x =>
-              val condition = if (e.filter.isDefined) {
-                e.filter.map(distinctAggFilterAttrLookup.get(_)).get
+              val condition = e.filter.map(distinctAggFilterAttrLookup.get(_)).getOrElse(None)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala:300]: nit: is `.getOrElse(None)` needed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 4]
```diff
@@ -293,12 +295,16 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
           val operators = expressions.map { e =>
             val af = e.aggregateFunction
             val naf = patchAggregateFunctionChildren(af) { x =>
-              val condition = if (e.filter.isDefined) {
-                e.filter.map(distinctAggFilterAttrLookup.get(_)).get
+              val condition = e.filter.map(distinctAggFilterAttrLookup.get(_)).flatten
+              if (distinctAggGroupLookup(e).contains(x)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala:299]: I'm wondering if we can simplify the logic a bit. The goal is to only do the replacement for the first child if all the children are foldable. How about
```
val af = e.aggregateFunction
if (af.children.forall(_.foldable)) {
  val firstChild = evalWithinGroup(id, af.children.head, condition)
  af.withNewChildren(firstChild +: af.children.drop(1)).asInstanceOf[AggregateFunction]
} else {
  // the original code
}
```

Then we don't need to change a lot of code to create `distinctAggGroupLookup`

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala:299]: OK

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 5]
```diff
@@ -292,13 +292,14 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
...
           val operators = expressions.map { e =>
             val af = e.aggregateFunction
-            val naf = patchAggregateFunctionChildren(af) { x =>
-              val condition = if (e.filter.isDefined) {
-                e.filter.map(distinctAggFilterAttrLookup.get(_)).get
-              } else {
-                None
+            val condition = e.filter.map(distinctAggFilterAttrLookup.get(_)).flatten
+            val naf = if (af.children.forall(_.foldable)) {
+              val firstChild = evalWithinGroup(id, af.children.head, condition)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala:297]: can we add some comments to explain why we are doing it?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala:300]: OK

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by cloud-fan** [comment]
github action passed, merging to master, thanks!

**Comment by beliefer** [comment]
@cloud-fan Thanks

**Pull request closed without merging**